lib/dependencies.js

```diff
@@ -0,0 +1,380 @@
+        depsForRoot.add(path.relative(depRoot, deps[i]));
+
+        // Create an inverse map so that when a dependency is invalidated
+        // we can track it back to the file that should be processed again.
+        let dependents = this.dependentsMap.get(deps[i]);
+        if (!dependents) {
+          dependents = [];
+          this.dependentsMap.set(deps[i], dependents);
+        }
+        dependents.push(referer);
+      }
+    });
+    return this;
+  }
+
+  _getDepsForRoot(dir) {
+    let depsForRoot = this.allDependencies.get(dir);
+    if (!depsForRoot) {
+      depsForRoot = new Set();
+      this.allDependencies.set(dir, depsForRoot);
+    }
+    return depsForRoot;
+  }
+
+  unseal() {
+    this.sealed = false;
+    this.allDependencies.clear();
+    this.fsTrees.clear();
+  }
+
+  countAll() {
+    let num = 0;
+    this.dependencyMap.forEach((deps) => {
+      num += deps.length;
+    });
+    return num;
+  }
+
+  /**
+   * Counts the number of unique dependencies.
+   *
+   * @returns {number}
+   */
+  countUnique() {
+    if (!this.sealed) {
+      throw new Error("Cannot count dependencies until after sealing them.");
+    } else {
+      return this.dependentsMap.size;
+    }
+  }
+
+  /**
+   * Set the dependencies for the file specified by `filePath`.
+   *
+   * @param filePath {string} relative path of the file that has dependencies.
+   * @param dependencies {Array<string>} absolute or relative paths the file
+   *   depends on. Relative paths are resolved relative to the directory
+   *   containing the file that depends on them.
+   */
+  setDependencies(filePath, dependencies) {
+    if (this.sealed) {
+      throw new Error("Cannot set dependencies when sealed");
+    }
+    /** @type {Array<string>} */
+    let absoluteDeps = [];
+    let fileDir = path.dirname(filePath);
+    for (let i = 0; i < dependencies.length; i++) {
+      let depPath = dependencies[i];
+      if (!path.isAbsolute(depPath)) {
+        depPath = path.normalize(path.resolve(this.rootDir, fileDir, depPath));
+      }
+      absoluteDeps.push(depPath);
+    }
+    this.dependencyMap.set(filePath, absoluteDeps);
+  }
+
+  /**
+   * Return a new, unsealed Dependencies that includes all the files and their
+   * dependencies except for the files provided (and their dependencies) are
+   * omitted.
+   *
+   * Note: this doesn't include the stat entries for the existing dependencies.
+   *
+   * @param files {Array<string>}
+   * @returns {Dependencies}
+   */
+  copyWithout(files) {
+    let newDeps = new Dependencies(this.rootDir);
+    for (let file of this.dependencyMap.keys()) {
+      if (!files.includes(file)) {
+        newDeps.setDependencies(file, this.dependencyMap.get(file));
+      }
+    }
+    return newDeps;
+  }
+
+  /**
+   * Get the dependency state and save it.
+   * Dependencies must be sealed.
+   * @returns {this}
+   */
+  captureDependencyState() {
+    this.fsTrees = this.getDependencyState();
```

Comment:
My definition of "sealed" in this context is that no new dependencies can be added so that the object can cache information in its internal state that assumes that dependencies won't change. Maybe a different term would align better with this definition?

Reply:
Oh I see, so "seal" is only sealing the portion of the object related to dependencies. That then seems fine.